marlowe-cli/src/Language/Marlowe/CLI/Transaction.hs

```diff
@@ -355,44 +439,62 @@ buildMinting connection signingKeyFile tokenNames metadataFile count expires lov
       (script, scriptHash) = mintingScript verification expires
       policy = PolicyId scriptHash
-      minting =
-        valueFromList
-          [
-            (
-              AssetId policy (AssetName $ fromBuiltin name)
-            , Quantity count
-            )
-          |
-            TokenName name <- tokenNames
-          ]
+      minting = tokenDistribution <&> \(TokenName name, count, recipient) -> do
+        let
+          value = valueFromList . pure $
+              (AssetId policy (AssetName $ fromBuiltin name) , Quantity count)
+        (fromMaybe changeAddress recipient, value)
+
       mintValue =
-        TxMintValue (toMultiAssetSupportedInEra era) minting
+        TxMintValue (toMultiAssetSupportedInEra era) (foldMap snd minting)
           . BuildTxWith
           . M.singleton policy
           $ SimpleScriptWitness (toSimpleScriptV2LanguageInEra era) SimpleScriptV2 (SScript script)
+
     metadata' <-
-      case metadata of
-        Just (A.Object metadata'') -> fmap (TxMetadataInEra (toTxMetadataSupportedInEra era))
-                                        . liftCli
-                                        . metadataFromJson TxMetadataJsonNoSchema
-                                        . A.Object
-                                        . KeyMap.singleton "721"
-                                        . A.Object
-                                        . KeyMap.singleton (Aeson.Key.fromString . BS8.unpack $ serialiseToRawBytesHex policy)
-                                        $ A.Object metadata''
-        _                          -> pure TxMetadataNone
-    void
-      $ buildClean
-          connection
-          [signingKeyFile]
-          lovelace
-          changeAddress
-          ((0, ) <$> expires)
-          mintValue
-          metadata'
-          bodyFile
-          timeout
-    liftIO . putStrLn $ "PolicyID " <> show policy
+      case metadataProps of
+        Just metadataProps' -> fmap (TxMetadataInEra (toTxMetadataSupportedInEra era))
+                              . liftCli
+                              . metadataFromJson TxMetadataJsonNoSchema
+                              . A.Object
+                              . KeyMap.singleton "721"
+                              . A.Object
+                              . KeyMap.singleton (Aeson.Key.fromString . BS8.unpack $ serialiseToRawBytesHex policy)
+                              $ A.Object metadataProps'
+        _               -> pure TxMetadataNone
+
+
+    -- TODO: use sensible coin selection here.
+    utxos <-
+      fmap (M.toList . unUTxO)
+        . queryInEra connection
+        . QueryUTxO
+        . QueryUTxOByAddress
+        . S.singleton
+        $ toAddressAny' changeAddress
+
+    let
+      inputs = fst <$> utxos
+
+    outputs <- for minting \(address, mintedValue) ->
+        makeTxOut address Nothing (mintedValue <> lovelaceToValue lovelace) ReferenceScriptNone
```

Comment:
This code fails when minting negative values (i.e., "burning") because it puts a negative value in the output, instead of finding the appropriate positive value in the input, and just outputting the difference, if any.

Reply:
https://input-output.atlassian.net/browse/SCP-4457